Documentation/rfc/v3api.md

```diff
@@ -0,0 +1,459 @@
+
+
+## Protobuf Defined API
+
+``` protobuf
+syntax = "proto3";
+
+// Interface exported by the server.
+service etcd {
+  // Range gets the keys in the range from the store.
+  // Maybe rename it to GET?
+  rpc Range(RangeRequest) returns (RangeResponse) {}
+
+  // Put puts the given key into the store.
+  // A put request increases the index of the store,
+  // and generates one event in the event history.
+  rpc Put(PutRequest) returns (PutResponse) {}
+
+  // Delete deletes the given key from the store.
+  // A delete request increase the index of the store,
+  // and generates one event in the event history.
+  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+
+  // Tnx processes all the requests in one transaction.
+  // A tnx request increases the index of the store,
+  // and generates one event in the event history.
+  rpc Tnx(TnxRequest) returns (TnxResponse) {}
+
+  // Watch watches the events happening or happened in etcd. Both input and output
+  // are stream. One watch rpc can watch for multiple ranges and get a stream of
+  // events. The whole events history can be watched unless compacted.
+  rpc Watch(stream WatchRequest) returns (stream WatchResponse) {}
+
+  // Compact compacts the event history in etcd. User should compact the
+  // event history periodically, or it will grow infinitely.
+  rpc Compact(CompactRequest) returns (CompactResponse) {}
+
+  // LeaseCreate creates a lease. A lease has a TTL. The lease will expire if the
+  // server does not receive a keepAlive within TTL from the lease holder.
+  // All keys attached to the lease will be expired and deleted if the lease expires.
+  // The key expiration generates an event in event history.
+  rpc LeaseCreate(LeaseCreateRequest) returns (LeaseCreateResponse) {}
+
+  // LeaseRevoke revokes a lease. All the key attached to the lease will be expired and deleted.
+  rpc LeaseRevoke(LeaseRevokeRequest) returns (LeaseRevokeResponse) {}
+
+  // LeaseAttach attaches keys with a lease.
+  rpc LeaseAttach(LeaseAttachRequest) returns (LeaseAttachResponse) {}
+
+  // LeaseTnx likes Tnx. It does one additional thing that all the keys put into the
+  // store are attached with the given lease. This is rpc is useful when you want to
+  // put a key and attach it to a lease atomically.
+  rpc LeaseTnx(LeaseTnxRequest) returns (LeaseTnxResponse) {}
+
+  // KeepAlive keeps the lease alive.
+  rpc LeaseKeepAlive(stream LeaseKeepAliveRequest) returns (stream LeaseKeepAliveResponse) {}
+}
+
+message RequestHeader {
+  // the cluster id to check with
+  // if the cluster id is not matched, an error is returned.
+  optional uint64 cluster_id = 1;
+}
+
+message ResponseHeader {
+  // an error type message?
+  optional string error = 1;
+  optional uint64 cluster_id = 2;
+  optional uint64 member_id = 3;
+  // index of the store when the requested was processed.
+  optional int64 index = 4;
+  optional uint64 raft_term = 5;
+}
+
+message RangeRequest {
```

Comment:
> On Apr 14, 2015, at 12:21 PM, Xiang Li <notifications@github.com> wrote:
> 
> In Documentation/rfc/v3api.md:
> 
> > +  // the cluster id to check with
> > +  // if the cluster id is not matched, an error is returned.
> > +  optional uint64 cluster_id = 1;
> > +}
> > +
> > +message ResponseHeader {
> > +  // an error type message?
> > +  optional string error = 1;
> > +  optional uint64 cluster_id = 2;
> > +  optional uint64 member_id = 3;
> > +  // index of the store when the requested was processed.
> > +  optional int64 index = 4;
> > +  optional uint64 raft_term = 5;
> > +}
> > +
> > +message RangeRequest {
> Do you anticipate/expect there to be delete tombstones in the keyspace?
> 
> The tombstones is in the event history. Do a range query with max_index. Fill in the gaps via another range query over the history from the max_index to now.
> 
> Does that make sense to you
> 
Yes, I think so.
> We can do the consistent query too. But it would make the API a little bit complicated. For each range query, we will return you a range ID and a timeout. Within the timeout, you can get consistent snapshot.
> 
I don't want to overcomplicate the store or API, certainly.  I *think* consistent paged range queries are valuable on their own to a wide range of users, as long as it does not break the core consistency guarantees of the store.  But I may also be bringing my own assumptions to the discussion.  If easy to emulate above, or can lead to complex problems on failover, it doesn't need to be in the API.
> —
> Reply to this email directly or view it on GitHub.
>